### PR TITLE
Optionally capture a dump during generational aware analysis

### PR DESCRIPTION
**Pull request by cshung**
This change introduces an extra option to allow generational aware analysis to capture a dump in addition to the trace files.

The trace file contains information about the type of object that held the young objects alive, but very often, we also need to know the exact instance. This is particularly useful when the young objects are leaked by a delegate, where we would like to know which callback is causing the problem.

@dotnet/dotnet-diag

### CODE REVIEW

**Review by noahfalk — APPROVED** [review]
Looks fine to me, @mikem8361 you want to take a look as well?

File: src/coreclr/vm/gcenv.ee.cpp [hunk 1]
```diff
@@ -1651,6 +1651,19 @@ void GCToEEInterface::AnalyzeSurvivorsFinished(size_t gcIndex, int condemnedGene
             s_forcedGCInProgress = false;
             reportGenerationBounds();
             FireEtwGenAwareEnd((int)gcIndex, GetClrInstanceId());
+            if (gcGenAnalysisDump)
+            {
+                EX_TRY
+                {
+#ifdef HOST_WIN32
```
Comment (Maoni0) [inline src/coreclr/vm/gcenv.ee.cpp:1658]: I have not seen this define used in the runtime code except in one file that I believe came from mono folks. the convention in the coreclr runtime code is to use 

```
#ifdef TARGET_UNIX
    PAL_Call();
#else 
    Windows_Call();
#endif //TARGET_UNIX
```
probably better to go with the convention.

File: src/coreclr/vm/gcenv.ee.cpp [hunk 2]
```diff
@@ -1651,6 +1651,19 @@ void GCToEEInterface::AnalyzeSurvivorsFinished(size_t gcIndex, int condemnedGene
...
+                    GenerateCrashDump (GENAWARE_DUMP_FILE_NAME, 2, false);
+#else
+                    PAL_GenerateCoreDump (GENAWARE_DUMP_FILE_NAME, 2, false);
```
Comment (Maoni0) [inline src/coreclr/vm/gcenv.ee.cpp:1661]: do we need to still capture a trace in this case seeing that we are already capturing a dump? you can load a dump in perfview and the genaware analysis would show up..

Comment (cshung) [reply src/coreclr/vm/gcenv.ee.cpp]: I made collecting a trace optional. There are a few reasons behind the design choice I made:

1. `COMPlus_GCGenAnalysisTrace` has 1 as a default - this is to maintain backward compatibility - users are expected to get a trace when both `GCGenAnalysisGen` and `GCGenAnalysisBytes` are set.

2. I make it possible to obtain both a trace and a dump at the same time. This is for the case of a Linux dump, where we might not be able to use the GenAware views in PerfView.

**Comment by cshung** [comment]
@Maoni0, can you please approve the PR if the change is good from the GC's perspective?
@noahfalk, can you please take a look again? I made some refactoring to the underlying code since your last approval. The refactoring should make it easier to capture a dump without having to have an ifdef at the call site.

**Review by Maoni0 — APPROVED** [review]

**Comment by Maoni0** [comment]
looks good from GC's POV.

**Review by noahfalk — APPROVED** [review]
Still looks good to me : )
(And thanks for the little refactor on GenerateDump)

**Review by hoyosjs — APPROVED** [review]
Dump changes and EP changes look like a cleanup to me :)

**Pull request merged**